spacy/kb_new/kb_interface.py

```diff
@@ -0,0 +1,204 @@
+"""Interface for KnowledgeBase for entity or concept linking."""
+import abc
+from typing import List, Any, Iterable, Iterator, Collection
+from spacy.kb import Candidate
+from spacy.tokens import Span
+from spacy.util import SimpleFrozenList
+
+
+class KnowledgeBase(abc.ABC):
+    """A KnowledgeBase implements everything necessary in order to leverage a specific data structure for (1) adding or
+    modifying existing entities and (2) looking up candidate entities for given mentions.
+    A "data structure" can be pretty much anything - custom or third-party, in-memory or on-disk or only available via
+    network, relational DB or graph DB or text search engine - as long as the necessary interface is fully implemented.
+    It is also possible to implement a data structure directly in a KnowledgeBase class.
+
+    DOCS: https://spacy.io/api/kb.
+    """
+
+    @abc.abstractmethod
+    @property
+    def entity_vector_length(self) -> int:
+        """Returns entity vector length.
+        RETURNS (int): Entity vector length.
+        """
+
+    @abc.abstractmethod
+    def add_vector(self, entity_vector: List[float]) -> Any:
+        """Add an entity vector to the vectors table.
+        entity_vector (List[float]): Entity vector to add.
+        RETURNS (int): Vector index.
+        """
+
+    @abc.abstractmethod
+    def add_entity(
+        self, entity_hash: int, freq: float, vector_index: Any, feats_row: int
+    ) -> Any:
+        """Add an entry to the vector of entries.
+        entity_hash (int): Hash for entity to add.
+        freq (float): Entity frequency.
+        vector_index (Any): Index of entity vector.
+        RETURNS (Any): Entity index.
+        """
+
+    @abc.abstractmethod
+    def add_aliases(
+        self, alias_hash: int, entry_indices: List[Any], probs: List[float]
+    ) -> Any:
+        """Connect a mention to a list of potential entities with their prior probabilities .
+        After calling this method, make sure to update also the _alias_index using the return value.
+        alias_hash (int): Hash for alias to add.
+        entry_indices (List[Any]): List of indices of entities to add.
+        RETURNS (Any): Alias index.
+        """
```

Comment:
I generally agree. Don't you think that adding entities to the store is something every `KnowledgeBase` should support?
Maybe not, considering the case of existing/external stores we just connect to...

After doing away with `CandidateSelector` and `kb_interface.KnowledgeBase` maybe it'd make sense to have `KnowledgeBase -> WritableKnowledgeBase` or something similar. I.e., `KnowledgeBase` would be the minimal interface required by `EntityLinker` (basically just looking up candidates) and `WritableKnowledgeBase` adds methods like `add_entity()` manipulating the data in the store.

Reply:
After giving this another thought, you are completely right. I was kinda fixated on mandating some structure for creating knowledge bases. Since this is done completely manual / outside of any spaCy-provided scaffold anyway, it would be quite pointless to do so. 

Instead I removed all methods not called from the current `EntityLinker` implementation (and modified them to their batch versions, where appropriate. I'd only go with batch calls for the new KB). 
I also added a new `Candidate`, since IMO the old one was implementation-specific as well due to its reliance and requirement of hashes for entity and alias. The new one accepts only the names/IDs. If that is too much of a limitation/performance hit, we could also allow passing either a "lookup ID" or the name itself and consider that in the getters.

Necessary changes to `EntityLinker` should be rather limited: we'd have call the batched methods instead of the single-item ones and extend the config to include `candidate_batch_size``or similar.